### PR TITLE
Introduce rebalance simulate tool

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

#402 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The tool leverages standalone ZK, Helix controller which run locally to simulate the rebalance process.
The tool also initializes mock participant threads for simulation state transitions.
After each operation, the tool will print the current cluster status and all state transitions happen after the previous print out.

### Tests

- [x] The following tests are written for this issue:

NA

- [ ] The following is the result of the "mvn test" command on the appropriate module:

(Copy & paste the result of "mvn test")

### Commits

- [ ] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

https://github.com/apache/helix/wiki/Weight-aware-Globally-Evenly-distributed-Rebalancer

### Code Quality

- [ ] My diff has been formatted using helix-style.xml

### CODE REVIEW

**Comment by jiajunwang** [comment]
To the reviewers, this is only for testing and evaluation. No logic should impact business logic.
And the tool will be continuously modified as we have different simulation requirement.

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddResource.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    new ConfigAccessor(zkClient).setResourceConfig(clusterName, _resourceName, resourceConfig);
+    admin.rebalance(clusterName, _resourceName,
+        is.getReplicaCount(admin.getInstancesInCluster(clusterName).size()));
+    return true;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Adding new resource: " + _resourceName;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddResource.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                new HashSet<String>(_accessor.getChildNames(_keyBuilder.liveInstances()));
+            return actualLiveNodes.contains(_instanceConfig.getInstanceName());
+          }
+        };
+    return verifier.verify(5000);
+  }
+
+  @Override
+  public String getDescription() {
+    return "Adding new node: " + _instanceConfig.getInstanceName();
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RebootNode.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                new HashSet<String>(_accessor.getChildNames(_keyBuilder.liveInstances()));
+            return actualLiveNodes.contains(_instanceName);
+          }
+        };
+    return verifier.verify(5000);
+  }
+
+  @Override
+  public String getDescription() {
+    return "Reboot node: " + _instanceName;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RebootNode.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RemoveNode.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      admin.getInstanceConfig(clusterName, _instanceName);
+      return false;
+    } catch (HelixException ex) {
+      return true;
+    }
+  }
+
+  @Override
+  public String getDescription() {
+    return "Remove node: " + _instanceName;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RemoveNode.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RemoveResource.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  @Override
+  public boolean execute(HelixAdmin admin, HelixZkClient zkClient, String clusterName) {
+    admin.dropResource(clusterName, _resourceName);
+    return admin.getResourceIdealState(clusterName, _resourceName) == null;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Remove resource: " + _resourceName;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/RemoveResource.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/Scripts.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            InstanceConfig instanceConfig = admin.getInstanceConfig(clusterName, node);
+            instanceConfig.setInstanceEnabled(true);
+            admin.setInstanceConfig(clusterName, node, instanceConfig);
+          }
+          return true;
+        }
+
+        @Override
+        public String getDescription() {
+          return "Re-enable all nodes in zone: " + zone;
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/Scripts.java]: Nit: String format like how you do it elsewhere in other Operations

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+  public void simulate(String zkServers, String clusterName, MetadataOverwrites overwrites,
+      List<Operation> operations, boolean shutdownAfterSimulate) throws Exception {
+    startZookeeper(zkServers);
+
+    _zkServers = zkServers;
+    _admin = new ZKHelixAdmin(zkServers);
+    _clusterName = clusterName;
+
+    HelixZkClient.ZkClientConfig clientConfig = new HelixZkClient.ZkClientConfig();
+    clientConfig.setZkSerializer(new ZNRecordSerializer()).setConnectInitTimeout(3 * 1000);
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: Magic number into a constant?

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+  private void stopZookeeper() {
+    if (_server != null) {
+      _server.shutdown();
+    }
+  }
+
+  /**
+   * The mock participant process to simulate a live Helix participant.
+   */
+  private class SimulateParticipantProcess {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: Could this be in a separate class? This Simulator is getting too big?

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      _simulator.getNodeToZoneMap()
+          .put(_instanceConfig.getInstanceName(), _instanceConfig.getZoneId());
+      _simulator.startNewProcess(_instanceConfig.getInstanceName(),
+          Arrays.stream(BuiltInStateModelDefinitions.values())
+              .map(modelDef -> modelDef.getStateModelDefinition().getId())
+              .collect(Collectors.toList()));
+    } catch (Exception e) {
+      return false;
+    }
+    ZkHelixClusterVerifier verifier =
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java]: Why is verifier only being used for AddNode?

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+
+/**
+ * The abstract simulator class that contains basic logic.
+ */
+public abstract class AbstractSimulator {
+  private final int DEFAULT_WAIT_TIME = 120000;
+  private final boolean _verbose;
+  private final int _transitionDelay;
+
+  private Map<String, SimulateParticipantProcess> _processMap = new HashMap<>();
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: Would `participantProcessMap` be a better name here? I got confused in the middle.

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+  private void executeScript(List<Operation> operations) throws IOException {
+    for (Operation operation : operations) {
+      echo("Start operation: " + operation.getDescription());
+      // Reset the count before operation.
+      for (SimulateParticipantProcess process : _processMap.values()) {
+        process.getAndResetTransitions();
+      }
+      if (operation.execute(_admin, _zkClient, _clusterName)) {
+
+        ClusterStateVerifier.verifyByPolling(() -> {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: Just add some comments here that we want to see at least 1 transition happening...

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+          for (SimulateParticipantProcess process : _processMap.values()) {
+            if (process.getTransitionCount() > 0) {
+              return true;
+            }
+          }
+          return false;
+        }, DEFAULT_WAIT_TIME);
+
+        if (!isClusterConverged()) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: An operation could trigger multiple transitions, right? What if multiple Participants got transitions but haven't fully completed them? Then would you want to break here right away?

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+    echo(desc);
+
+    ConfigAccessor configAccessor = new ConfigAccessor(_zkClient);
+    ClusterConfig clusterConfig = configAccessor.getClusterConfig(_clusterName);
+    List<String> capacityKeys = clusterConfig.getInstanceCapacityKeys();
+
+    Set<String> resources = new HashSet<>(_admin.getResourcesInCluster(_clusterName));
+
+    Map<String, Integer> partitionCount = new HashMap<>();
+    Map<String, Map<String, Integer>> perStatePartitionCount = new HashMap<>();
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: Add comments on what each element in the map represents?

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/MetadataOverwrites.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.helix.model.ClusterConfig;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.ResourceConfig;
+
+/**
+ * Overwrites the cluster metadata for the simulation purposes.
+ */
+public interface MetadataOverwrites {
```
[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/MetadataOverwrites.java]: I want to ensure the user understands that this class if for overwriting the existing metadata which is copied from somewhere else. So strictly speaking, it is not a copy tool.

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java (hunk 9 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/operations/AddNode.java]: This is to ensure the participant process starts and register correctly.

File: helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java (hunk 12 above)
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/examples/rebalancer/simulator/AbstractSimulator.java]: I tried to add some more validations to ensure the converged assignment is a valid one. This should be good for the simulate. But still not good enough for our integration tests.

**Comment by jiajunwang** [comment]
Update for generating the report, I will refine the format later.

**Comment by jiajunwang** [comment]
This tool won't be required for now. I'm closing this PR and will leave the tool in my branch.

**Pull request closed without merging**